hare/broker.go

```diff
@@ -0,0 +1,103 @@
+package hare
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/spacemeshos/go-spacemesh/hare/pb"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"hash/fnv"
+	"sync"
+)
+
+const InboxCapacity = 100
+
+// Stopper is used to add stoppability to an object
+type Stopper struct {
+	count   uint8         // the number of listeners
+	channel chan struct{} // listeners listen to this channel
+}
+
+func NewStopper(listenersCount uint8) *Stopper {
+	return &Stopper{listenersCount, make(chan struct{})}
+}
+
+// Stops all listening instances (should be called only once)
+func (stopper *Stopper) Stop() {
+	for i := uint8(0); i < stopper.count; i++ {
```

Comment:
you can use the close(ch) func to signal all listeners on this channel that it has closed